src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -160,16 +166,214 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+                {
+                    return minimumSegmentSize;
+                }
+            }
+
+            // cannot find an appropriate size for segments to be combined.
+            return -1;
+        }
+
+        /// <summary>
+        /// Combines continguous segments with lengths that are each less than or equal to the minimum segment size.
+        /// </summary>
+        private static void CombineSegments(ArrayBuilder<SourceText> segments, int minimumSegmentSize)
+        {
+            for (int i = 0; i < segments.Count - 1; i++)
+            {
+                if (segments[i].Length <= minimumSegmentSize)
+                {
+                    int combinedLength = segments[i].Length;
+
+                    // count how many contiguous segments are reducible
+                    int count = 1;
+                    for (int j = i + 1; j < segments.Count; j++)
+                    {
+                        if (segments[j].Length <= minimumSegmentSize)
+                        {
+                            count++;
+                            combinedLength += segments[j].Length;
+                        }
+                    }
+
+                    // if we've got at least two, then combine them into a single text
+                    if (count > 1)
+                    {
+                        var encoding = segments[i].Encoding;
+                        var algorithm = segments[i].ChecksumAlgorithm;
+
+                        var writer = SourceTextWriter.Create(encoding, algorithm, combinedLength);
+
+                        while (count > 0)
+                        {
+                            segments[i].Write(writer);
+                            segments.RemoveAt(i);
+                            count--;
+                        }
+
+                        var newText = writer.ToSourceText();
+
+                        segments.Insert(i, newText);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Determines the total number of segments that would remain if the contiguous segments of length less than or equal to
+        /// the minimum segment size are combined together.
+        /// </summary>
+        private static int GetReducedSize(ArrayBuilder<SourceText> segments, int minimumSegmentSize)
+        {
+            int numberOfSegmentsReduced = 0;
+
+            for (int i = 0; i < segments.Count - 1; i++)
+            {
+                if (segments[i].Length <= minimumSegmentSize)
+                {
+                    // count how many contiguous segments are reducible
+                    int count = 1;
+                    for (int j = i + 1; j < segments.Count; j++)
+                    {
+                        if (segments[j].Length <= minimumSegmentSize)
+                        {
+                            count++;
+                        }
+                    }
+
+                    if (count > 1)
```

Comment:
Observation: if you remove the I++ from the outermost loop, then this if condition goes away. I think that makes the algorithm easier (or at least makes me sleep better because there's less branching.)